sympy/solvers/solvers.py

```diff
@@ -2747,7 +2747,7 @@ def equal(expr1, expr2):
             elif lhs.func == LambertW:
                 return _vsolve(lhs.args[0] - rhs*exp(rhs), sym, **flags)
 
-        rewrite = lhs.rewrite(exp)
+        rewrite = lhs.rewrite([TrigonometricFunction, HyperbolicFunction], exp, evaluate=True)
         if rewrite != lhs:
             return _vsolve(rewrite - rhs, sym, **flags)
```

Comment:
This approach of recursing if rewrite changes something is inherently fragile. If you trace through the calls from `_vsolve` back to here how can you ensure that passing `rewrite - lhs` will not just lead to this function receiving the same `eq` that it started with?

Reply:
I agree. This change makes it slightly less fragile (and fixes the one known issue), but if anything ever rewrites the trig functions back to exponentials, we're back where we started. Ideally solve would either not be written recursively at all, or be written in such a way that infinite loops/recursions are impossible. 

I think solve (or solveset) should be structured in a way that is much more explicit about the steps involved for solving an equation. There are three core steps involved in solving an equation

- Matching: similar to the matching step in the ODE solvers. Trying to figure out from the input which of many solvers should be attempted. This can also just be considered an optimization since you can just try the solvers in order til you find one that works, assuming they can never produce a wrong answer on invalid input. 
- Rewriting: trying various rewrites to put an expression into a form fit for a given solver. Only some rewrites should be tried for a given equation and for a given solver. Simplification is also a form of rewriting.
- Decomposition: writing f(x) as g(h(x)) so that g and h can be inverted separately. 

And then at the core there are specific "inversion" solvers, similar to the ODE solvers (e.g., a solver that can invert polynomials, a solver that knows how to produce LambertW-type solutions, etc.).

These should be split out into separate routines so that a central algorithm can apply them. dsolve is a decent model of how this sort of thing can be done. dsolve() can solve way more "types" of equations than solve(), and it's clear that that sort of approach can scale (although dsolve doesn't yet handle rewriting and decomposition very well). 

Right now matching is done in an ad hoc manner, and rewriting and recursion are done naively using recursion. But that means that everything is muddled together. It also means that there's no central algorithm that knows what is going on. That's why you get recursion errors, because each step only knows its own step of the process, and doesn't know about what has already been tried. 

solveset is a little cleaner in how its written, but it still doesn't really use this design.

Reply:
It's okay to use recursion provided you know that you are recursing onto something that will not loop back so it should be e.g. something that is smaller. There is nothing wrong with recursing say from an expression to call an operation on its args because they are necessarily "smaller" and therefore cannot loop back.

The rewrite here is okay but it should be done before entering the recursive part of solve. That's what happens with hyperbolic functions:
https://github.com/sympy/sympy/blob/68803e50d9d76a8b14eaf2a537f90cb86d0a87be/sympy/solvers/solvers.py#L965-L968

The problem here is that rewrite produces something equivalent to what we started with and the rest of the solve code deliberately tries to canonicalise equivalent things. If you condense the recursion here to a single function it looks basically like this:
```python
def _solve(eq_orig):
    # transform all equivalent equations to a canonical form
    eq_canon = canonicalise(eq_orig)

   # do lots of stuff and return if it works.
   # if we get to the lines below then the stuff didn't work

   # change to a different but equivalent equation
   eq_changed = eq_canon.rewrite()
   if eq_changed != eq_canon:
       return _solve(eq_changed)
```
It's obvious that this can lead to infinite recursion which is why we have `if eq_changed != eq_canon` but that is not a robust guard. The *minimal* check to guard against infinite recursion should be `if canonicalise(eq_changed) != eq_canon` but there's no way to check that because there is no callable `canonicalise` function (it's spread out over all of `solve`).

Reply:
I'm not sure if there is a suggestion for a change here. 

The problem with your general idea is you cannot extract `canonicalize()` into a single operation. It happens all over the codebase, not just in solve. For this particular issue, the canonicalization happens in Poly. But it really could happen anywhere because it's the result of rewriting an unevaluated expression back into its evaluated form. The motivation in #24549 is to not do rewrites that produce unevaluated expressions, so at least that level of basic re-canonicalization won't break this logic.

Reply:
> The problem with your general idea is you cannot extract `canonicalize()` into a single operation

I wan't proposing that as a solution but just as a demonstration of why the implicit assumption in this recursive call is broken.

To be clear I don't object to the change highlighted in the diff above which can reduce the problem in some cases. I just don't consider it to be a proper fix because the root problem with this code is still there.

Reply:
I think it would also be possible to a more complex recursion error where rewrites cycle back to each other. The only real way to fix this is to have a top-level solve algorithm that applies the rewrites and other heuristics and keeps track of what has been tried. And that design has other advantages too like making the algorithms more modular.